### PR TITLE
Update deployment to remove dependency on Azure PowerShell TokenCache

### PR DESCRIPTION
**Pull request by JamesDawson**
Utilises `az rest` to make requests to the Azure Graph API to negate the need for direct use of the access token.

Also took the opportunity to simplify the AppRole Assignment functionality and make it more robust.

### CODE REVIEW

File: Solutions/Marain.Instance.Deployment/Deploy-MarainInstanceInfrastructure.ps1 [hunk 1]
```diff
@@ -1,958 +1,1092 @@
...
+        [hashtable] $Headers = @{}
+    )
+
+    # Ensure we always have the 'Content-Type' header
+    if ( !$Headers.ContainsKey("Content-Type") ) {
+        $Headers += @{ "Content-Type" = "application/json" }
+    }
+
+    if (@("GET", "DELETE") -contains $Method) {
+        $uriEscaped = $Uri.Replace("'", "''")
```
Comment (MikeEvansLarah) [inline Solutions/Marain.Instance.Deployment/Deploy-MarainInstanceInfrastructure.ps1]: How come you don't have to escape the URI for PATCH/POST/PUT requests?

File: Solutions/Marain.Instance.Deployment/Deploy-MarainInstanceInfrastructure.ps1 [hunk 2]
```diff
@@ -1,958 +1,1092 @@
...
+        $InstanceDeploymentContext.MarainCliPath += '.exe'
+    } 
+    if ( !(Test-Path $InstanceDeploymentContext.MarainCliPath)) {
+        & dotnet tool install -g $marainGlobalToolName --version $marainGlobalToolVersion
+    }
+
+    # Lookup the identity of the deployment user, as we need their objectId to grant keyvault access
+    $currentContext = Get-AzContext
+    if ($currentContext.Account.Id -imatch "@") {
+        $currentUser = Get-AzAdUser -UserPrincipalName $currentContext.Account.Id
```
Comment (MikeEvansLarah) [inline Solutions/Marain.Instance.Deployment/Deploy-MarainInstanceInfrastructure.ps1]: Will this fail if we are running without graph permissions?

**Review by MikeEvansLarah — CHANGES_REQUESTED** [review]

File: azure-pipelines.yml [hunk 3]
```diff
@@ -1,120 +1,120 @@
...
-              Get-AzADApplication -DisplayName $appName | Select -First 1 | Remove-AzADApplication -Verbose -Force
-          }
-        # Temporarily force a downgrade of the AzurePowerShell cmdlets because (as of September 2020 and v4.6.0) a bug
-        # was introduced which breaks use of securestring with ARM templates:
-        # https://github.com/Azure/azure-powershell/issues/12792
-        # https://github.com/Azure/azure-powershell/issues/12819
-        # Once this issue is fixed, we should revert back to using the latest version of the cmdlets.
-        preferredAzurePowerShellVersion: '4.4.0'
-        pwsh: true
+# Starter pipeline
```
Comment (MikeEvansLarah) [inline azure-pipelines.yml]: Not sure why this file is showing changes. Is it line endings? Maybe we should have a .gitattributes file to make line endings consistent across Windows/*nix

File: Solutions/Marain.Instance.Deployment/Deploy-MarainInstanceInfrastructure.ps1 (hunk 1 above)
Comment (JamesDawson) [reply Solutions/Marain.Instance.Deployment/Deploy-MarainInstanceInfrastructure.ps1]: This is lifted from what is in `corvus.deployment`, I'll add an issue in that repo.

File: azure-pipelines.yml (hunk 3 above)
Comment (JamesDawson) [reply azure-pipelines.yml]: Line endings renormalized

File: Solutions/Marain.Instance.Deployment/Deploy-MarainInstanceInfrastructure.ps1 (hunk 2 above)
Comment (JamesDawson) [reply Solutions/Marain.Instance.Deployment/Deploy-MarainInstanceInfrastructure.ps1]: A service principal has access to their own entry so it works without explicit graph permissions - I've not specifically tested a real user account without the default graph access they normally get, but I think it's sufficiently edge-case to leave it as-is for the moment.  Maybe we can review it again as part of the port to `corvus.deployment`.

**Review by MikeEvansLarah — APPROVED** [review]

**Pull request merged**